system-helper/flatpak-system-helper.c

```diff
@@ -193,6 +232,160 @@ no_progress_cb (OstreeAsyncProgress *progress, gpointer user_data)
+    {
+      if (g_strcmp0 (pull->user_folder_name, g_unix_mount_get_mount_path (l->data)) == 0)
+        {
+          flatpak_system_helper_complete_create_pull_cache_dir (pull->object,
+                                                                pull->invocation,
+                                                                pull->user_folder_name);
+	  break;
+        }
+    }
+  // TODO: Way to return error if mount isn't successful or timeout
+}
+
+static gboolean
+handle_create_pull_cache_dir (FlatpakSystemHelper   *object,
+                              GDBusMethodInvocation *invocation,
+                              const gchar           *arg_ref,
+                              const gchar           *optional_commit,
+                              const gchar           *arg_installation)
+{
+  GDBusConnection *connection = g_dbus_method_invocation_get_connection (invocation);
+  const gchar *sender = g_dbus_method_invocation_get_sender (invocation);
+  FlatpakSystemHelperOngoingPull *new_pull;
+  g_autoptr(FlatpakDir) system = NULL;
+  g_autoptr(GError) error = NULL;
+  g_autofree gchar *user_folder = NULL;
+  g_autofree gchar *root_folder = NULL;
+  g_autofree gchar *flatpak_dir = NULL;
+  g_autofree gchar *bindfs_path = NULL;
+  guint uid_pull_count = 0;
+  uid_t uid;
+
+  system = dir_get_system (arg_installation, &error);
+  if (system == NULL)
+    {
+      g_dbus_method_invocation_return_gerror (invocation, error);
+      return TRUE;
+    }
+
+  if (ongoing_pulls == NULL)
+    ongoing_pulls = g_ptr_array_new ();
+
+  for (guint i = 0; i < ongoing_pulls->len; i++)
+    {
+      FlatpakSystemHelperOngoingPull *pull_tmp;
+      pull_tmp = g_ptr_array_index (ongoing_pulls, i);
+      if (pull_tmp->uid == uid)
+        {
+          uid_pull_count++;
+          // TODO: If there is a already a partial pull available, return that.
+        }
+    }
+
+  uid = get_connection_uid (invocation, &error);
+  if (uid_pull_count > MAX_PULLS_PER_USER)
+    {
+      g_dbus_method_invocation_return_error (invocation, G_DBUS_ERROR, G_DBUS_ERROR_FAILED,
+                                             "Maximum number of pulls per user exceeded for UID: %u", uid);
+      return TRUE;
+    }
+
+  flatpak_dir = g_file_get_path (flatpak_dir_get_path (system));
+  root_folder = g_mkdtemp_full (g_build_filename (flatpak_dir, "repo", "tmp", "flatpak-cache-XXXXXX", NULL), 0755);
+  user_folder = g_strdup_printf ("%s-bindfs", root_folder);
+  if (g_mkdir_with_parents (user_folder, 0755) == -1)
+    {
+      g_dbus_method_invocation_return_error (invocation, G_DBUS_ERROR, G_DBUS_ERROR_FAILED,
+                                             "Cannot create directory for bindfs mount at: %s", user_folder);
+      return TRUE;
+    }
+
+  new_pull = system_helper_ongoing_pull_new (object, invocation, uid, arg_ref, optional_commit, user_folder);
+  g_ptr_array_add (ongoing_pulls, new_pull);
+
+  bindfs_path = g_find_program_in_path ("bindfs");
```

Comment:
This should be checked much earlier in this function.

Reply:
I have gone all the way to check `bindfs` in the client code. If `bindfs` is not found, then it fallbacks on the existing code path for the pull.

These bit is addressed in https://github.com/endlessm/flatpak/pull/115/commits/b55725c01a0eb1a016dde7e362052c02506d20fc

Reply:
Also, this(`g_find_program_in_path`) needs to be queried for `bindfs` path which is used for the spawing the subprocess below. Maybe get rid of the `if` condition here (as it's already being checked on the client side now)?